File: frontend/src/types/skill.ts
export type SkillLevel = 'BEGINNER' | 'INTERMEDIATE' | 'ADVANCED' | 'EXPERT'

export interface Skill {
  id: number
  name: string
}

export interface SkillCreate {
  name: string
}

export type SkillUpdate = Partial<SkillCreate>

export interface EmployeeSkill {
  skill_id: number
  skill_name: string
  level: SkillLevel
}

export interface EmployeeSkillCreate {
  skill_id: number
  level: SkillLevel
}

export interface EmployeeSkillUpdate {
  level: SkillLevel
}

export interface ProjectSkill {
  skill_id: number
  skill_name: string
  required_level: SkillLevel
}

export interface ProjectSkillCreate {
  skill_id: number
  required_level: SkillLevel
}

export interface ProjectSkillUpdate {
  required_level: SkillLevel
}

File: backend/app/schemas/skill.py
from typing import Annotated, Any

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, field_validator

from app.models.skill import SkillLevel


SkillName = Annotated[str, StringConstraints(strip_whitespace=True, min_length=1)]
PositiveId = Annotated[int, Field(gt=0)]


class SkillCreate(BaseModel):
    name: SkillName


class SkillUpdate(BaseModel):
    name: SkillName | None = None

    @field_validator("name", mode="before")
    @classmethod
    def name_cannot_be_null(cls, value: Any) -> Any:
        if value is None:
            raise ValueError("Field cannot be null")
        return value


class SkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str


class EmployeeSkillCreate(BaseModel):
    skill_id: PositiveId
    level: SkillLevel


class EmployeeSkillUpdate(BaseModel):
    level: SkillLevel


class EmployeeSkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    skill_id: int
    skill_name: str
    level: SkillLevel


class ProjectSkillCreate(BaseModel):
    skill_id: PositiveId
    required_level: SkillLevel


class ProjectSkillUpdate(BaseModel):
    required_level: SkillLevel


class ProjectSkillRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    skill_id: int
    skill_name: str
    required_level: SkillLevel
